engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/SafeHtmlValidator.java

```diff
@@ -56,10 +56,14 @@ public void initialize(SafeHtml safeHtmlAnnotation) {
 		whitelist.addTags( safeHtmlAnnotation.additionalTags() );
 
 		for ( SafeHtml.Tag tag : safeHtmlAnnotation.additionalTagsWithAttributes() ) {
-			whitelist.addAttributes( tag.name(), tag.attributes() );
-			if ( tag.protocols().length > 0 ) {
-				for ( String attribute : tag.attributes() ) {
-					whitelist.addProtocols( tag.name(), attribute, tag.protocols() );
+			whitelist.addTags( tag.name() );
+			if ( tag.attributes().length > 0 ) {
+				whitelist.addAttributes( tag.name(), tag.attributes() );
+			}
+			if ( tag.additionalAttributesWithProtocols().length > 0 ) {
+				for ( SafeHtml.Tag.Attribute attribute : tag.additionalAttributesWithProtocols() ) {
+					whitelist.addAttributes( tag.name(), attribute.name() );
+					whitelist.addProtocols( tag.name(), attribute.name(), attribute.protocols() );
```

Comment:
I'll double check this. But as I remember there are different Sets and Maps for each thing (tag, attribute, protocol etc) and validation will fail if you add a protocol for example `data` for `img` -> `src` then validation will fail as `img` tag is absent in tags, and the same for the attribute...
But again - I'll double check if that's really so :)